src/util/expr_cast.h

```diff
@@ -54,22 +56,39 @@ template <typename T, typename TExpr>
 optionalt<std::reference_wrapper<typename std::remove_reference<T>::type>>
 expr_try_dynamic_cast(TExpr &base)
 {
-  typedef typename std::decay<T>::type TUnderlying;
-  typedef typename std::remove_reference<T>::type TConst;
+  typedef typename std::decay<T>::type TUnderlyingt;
+  typedef typename std::remove_reference<T>::type TConstt;
   static_assert(
     std::is_same<typename std::remove_const<TExpr>::type, exprt>::value,
     "Tried to expr_try_dynamic_cast from something that wasn't an exprt");
   static_assert(
     std::is_reference<T>::value,
     "Tried to convert exprt & to non-reference type");
   static_assert(
-    std::is_base_of<exprt, TUnderlying>::value,
+    std::is_base_of<exprt, TUnderlyingt>::value,
     "The template argument T must be derived from exprt.");
-  if(!can_cast_expr<TUnderlying>(base))
+  if(!can_cast_expr<TUnderlyingt>(base))
     return {};
-  T value=static_cast<T>(base);
-  validate_expr(value);
-  return std::reference_wrapper<TConst>(value);
+  T ret=static_cast<T>(base);
+  validate_expr(ret);
+  return std::reference_wrapper<TConstt>(ret);
+}
+
+/// \brief Try to cast a reference to a generic exprt to a specific derived
+///    class
+/// \tparam T The reference or const reference type to \a TUnderlying to cast
+///    to
+/// \tparam TExpr The original type to cast from, either exprt or const exprt
+/// \param base Reference to a generic \ref exprt
+/// \return Reference to object of type \a TUnderlying
+///   or valueless optional if \a base is not an instance of \a TUnderlying
+template <typename T, typename TExpr>
+optionalt<std::reference_wrapper<typename std::remove_reference<T>::type>>
+expr_try_checked_cast(TExpr &base)
```

Comment:
This operation doesn't make sense. If you assert your check will pass then why would you force yourself to write code to handle the case that you're saying can't happen?